src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs

```diff
@@ -58,170 +54,45 @@ internal abstract class BDDAlgebra : IBooleanAlgebra<BDD>
         /// </summary>
         public BDDAlgebra() => _mintermGen = new MintermGenerator<BDD>(this);
 
-        /// <summary>
-        /// Assumes op is a binary commutative operation: one of OR, AND, XOR.
-        /// Treats the arguments as if they are unordered.
-        /// Orders left and right by hashcode in the constructed key.
-        /// </summary>
-        private static BoolOpKey CreateBoolOpKey(BoolOp op, BDD left, BDD right) =>
-            left.GetHashCode() <= right.GetHashCode() ?
-                new BoolOpKey(op, left, right) :
-                new BoolOpKey(op, right, left);
-
         /// <summary>
         /// Create a BDD with given ordinal and given one and zero child.
         /// Returns the BDD from the cache if it already exists.
-        /// Must be executed in a single thread mode.
         /// </summary>
-        public BDD GetOrCreateBDD(int ordinal, BDD? one, BDD? zero)
-        {
-            var key = new BDD(ordinal, one, zero);
-            if (!_bddCache.TryGetValue(key, out BDD? set))
-            {
-                set = key;
-                _bddCache.Add(set);
-            }
-
-            return set;
-        }
+        public BDD GetOrCreateBDD(int ordinal, BDD? one, BDD? zero) =>
+            _bddCache.GetOrAdd(new BDDKey(ordinal, one, zero), key => new BDD(ordinal, one, zero));
```

Comment:
This is going to incur closure/delegate allocations even for the fast path where the object already exists.  You can avoid that in a few ways, but the simplest here is probably to just use the overload that let's you pass in additional state, e.g.
```suggestion
        public BDD GetOrCreateBDD(int ordinal, BDD? one, BDD? zero) =>
            _bddCache.GetOrAdd(new BDDKey(ordinal, one, zero), (key, state) => new BDD(state.ordinal, state.one, state.zero), (ordinal, one, zero));
```

Reply:
Good to know, thanks! I've gotten too used to C++ where these would end up being stack allocations.